frontend/simple/js/transactions.js

```diff
@@ -0,0 +1,269 @@
+import sbp from '../../../shared/sbp'
+
+let _steps = new WeakMap()
+let _scope = new WeakMap()
+let _cursor = new WeakMap()
+let _description = new WeakMap()
+let _isRecoverable = new WeakMap()
+let _wait = new WeakMap()
+// Why would need this?  to make our retries grow in a common sense ways
+function fibSequence (num) {
+  let output = 0
+  let sequence = [ 0, 1 ]
+  for (var i = 0; i < num; i++) {
+    sequence.push(sequence[ 0 ] + sequence[ 1 ])
+    sequence.splice(0, 1)
+    output = sequence[ 1 ]
+  }
+  return output
+}
+
+let transactionDB = localforage.createInstance({ name: 'Transactions', storeName: 'Transactions' })
+export class Transaction {
+  constructor (description, recoverable) {
+    _description.set(this, description)
+    _steps.set(this, [])
+    _scope.set(this, {})
+    _isRecoverable.set(this, !!recoverable)
+    _cursor.set(this, 0)
+    let res
+    let rej
+    let promise = new Promise((resolve, reject) => (res = resolve) && (rej = reject))
+    _wait.set(this, {promise, resolve: res, reject: rej})
+  }
+  get isRecoverable () {
+    return _isRecoverable.get(this)
+  }
+  wait () {
+    let wait = _wait.get(this)
+    return wait.promise
+  }
+  addStep (step) {
+    let steps = _steps.get(this)
+    steps.push(step)
+  }
+  // This function creates variables in the transaction's scope to be referenced by steps
+  setInScope (name, value) {
+    let scope = _scope.get(this)
+    scope[name] = value
+    _scope.set(this, scope)
+  }
+  async run (serviceRegistry) {
+    let steps = _steps.get(this)
+    let scope = _scope.get(this)
+    let cursor = _cursor.get(this)
+    let wait = _wait.get(this)
+    for (let i = cursor; i < steps.length; i++) {
+      let args = {}
+      // Keep track of a cursor of what step in the transaction we are operating
+      _cursor.set(this, i)
+      let currentStep = steps[ i ]
+      // Build arguments from services and scoped data
+      for (let argPair of Object.entries(currentStep.args)) {
+        args[argPair[0]] = scope[argPair[1]]
+      }
+      args.setInScope = this.setInScope.bind(this)
+      // Attempt to run step
+      try {
+        await sbp(currentStep.execute, args)
+      } catch (ex) {
+        wait.reject(new Error(`Step - ${currentStep.description || ''} failed with error, \n ${ex}`))
+        throw ex
+      }
+    }
+    wait.resolve()
+  }
+  toJSON () {
+    return {
+      steps: _steps.get(this),
+      scope: _scope.get(this),
+      cursor: _cursor.get(this),
+      isRecoverable: _isRecoverable.get(this),
+      description: _description.get(this)
+    }
+  }
+  static isTransaction (obj) {
+    if (typeof obj !== 'object') return false
+    if (!Array.isArray(obj.steps)) return false
+    if (typeof obj.scope !== 'object') return false
+    if (!Number.isInteger(obj.cursor)) return false
+    if (typeof obj.isRecoverable !== 'boolean') return false
+    if (typeof obj.description !== 'string') return false
+    return true
```

Comment:
Let's discuss a different way of doing this that doesn't involve the need for this function

Reply:
`note-to-self:` this is a very heavy function (computationally) and so am hoping it turns out to either:

- not be necessary
- for there to be a different way of doing it that's less computationally heavy
- or for it to be run only in rare circumstances

Stylistically, I would personally prefer if it be written in a chain'd series of `&&`'s and a single return call.